python_modules/dagster/dagster/_streamline/asset_materialization_health.py

```diff
@@ -0,0 +1,176 @@
+
+
+@whitelist_for_serdes
+@record.record
+class AssetMaterializationHealthState:
+    """For tracking the materialization health of an asset, we only care about the most recent
+    completed materialization attempt for each asset/partition. This record keeps track of the
+    assets/partitions that are currently in a successful state and those that are in a failed state.
+    If an asset/partition is currently being materialized, it will not move to a new state until after
+    the materialization attempt is complete.
+
+    In the future, we may want to expand this to track the last N materialization successes/failures for
+    each asset. We could also maintain a list of in progress materializations, but that requires streamline to be
+    better able to handle runs being deleted.
+    """
+
+    materialized_subset: SerializableEntitySubset[AssetKey]
+    failed_subset: SerializableEntitySubset[AssetKey]
+    partitions_snap: Optional[PartitionsSnap]
+
+    @property
+    def partitions_def(self) -> Optional[PartitionsDefinition]:
+        if self.partitions_snap is None:
+            return None
+        return self.partitions_snap.get_partitions_definition()
+
+    @classmethod
+    def default(cls, asset_key: AssetKey) -> "AssetMaterializationHealthState":
+        return AssetMaterializationHealthState(
+            materialized_subset=SerializableEntitySubset(key=asset_key, value=False),
+            failed_subset=SerializableEntitySubset(key=asset_key, value=False),
+            partitions_snap=None,
+        )
+
+    @classmethod
+    async def compute_for_asset(
+        cls,
+        asset_key: AssetKey,
+        partitions_def: Optional[PartitionsDefinition],
+        loading_context: LoadingContext,
+    ) -> "AssetMaterializationHealthState":
+        if partitions_def is not None:
+            (
+                materialized_partition_subset,
+                failed_partition_subset,
+                _,
+            ) = get_partition_subsets(
+                loading_context.instance,
+                loading_context,
+                asset_key,
+                loading_context.instance,
+                partitions_def,
+            )
+
+            if materialized_partition_subset is None or failed_partition_subset is None:
+                check.failed("Expected partitions subset for a partitioned asset")
+
+            return cls(
+                materialized_subset=SerializableEntitySubset(
+                    key=asset_key, value=materialized_partition_subset
+                ),
+                failed_subset=SerializableEntitySubset(
+                    key=asset_key, value=failed_partition_subset
+                ),
+                partitions_snap=PartitionsSnap.from_def(partitions_def),
+            )
+
+        asset_record = await AssetRecord.gen(loading_context, asset_key)
+        if asset_record is None:
+            return cls.default(asset_key)
```

Comment:
Is this `default` used elsewhere? I can see why it always produces an unpartitioned record here, but it's a bit weird for it to be called something so general when it's really only the default for unpartitioned assets.